### PR TITLE
[1LP][RFR] Initial Commit: Added new functionality to Generate JDR Reports.

### PR DESCRIPTION
**Pull request by hhovsepy**
Added basic tests.
{{pytest: cfme/tests/middleware/test_jdr_report.py -v --use-provider hawkular}}

### CODE REVIEW

**Comment by hhovsepy** [comment]
upstream failures are expected

File: cfme/middleware/jdr_report.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
+import re
+
```
Comment (quarckster) [inline cfme/middleware/jdr_report.py]: This blank line is not needed.

File: cfme/middleware/jdr_report.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+def _db_select_query(server):
+    """column order: `id`, `queued_at`, `requesting_user`, `status`"""
+    t_ms = current_appliance.db.client['middleware_servers']
+    t_mdr = current_appliance.db.client['middleware_diagnostic_reports']
+    query = current_appliance.db.client.session.query(
+        t_mdr.id, t_mdr.queued_at,
+        t_mdr.requesting_user.label('username'),
+        t_mdr.status)\
+        .join(t_mdr, t_mdr.middleware_server_id == t_ms.id)
```
Comment (quarckster) [inline cfme/middleware/jdr_report.py]: **REQUIRED**
```python
query = current_appliance.db.client.session.query(
    t_mdr.id,
    t_mdr.queued_at,
    t_mdr.requesting_user.label('username'),
    t_mdr.status).join(t_mdr, t_mdr.middleware_server_id == t_ms.id)
```

File: cfme/middleware/jdr_report.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        size: size of the JDR Report file
+        queued_at: queued time of the JDR Report
+        db_id: database row id of JDR Report
+
+    Usage:
+
+        myreport = JDRReport(username='admin', status='Ready', server=eap_server)
+        myreport.download()
+
+        myreports = JDRReport.reports()
```
Comment (quarckster) [inline cfme/middleware/jdr_report.py]: **REQUIRED**
```python
Usage:

.. code-block:: python

    myreport = JDRReport(username='admin', status='Ready', server=eap_server)
    myreport.download()

    myreports = JDRReport.reports()
```

File: cfme/middleware/jdr_report.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        return False
+
+    @classmethod
+    def _queued_after(cls, report, date_after):
+        return datetime.strptime(
+            re.sub(r' \+\d+', '', report.queued_at), '%a, %d %b %Y %H:%M:%S') >= date_after
+
+    def delete(self):
+        """ Deletes the JDR Report """
+        # @TODO
```
Comment (quarckster) [inline cfme/middleware/jdr_report.py]: **REQUIRED**
```python
raise NotImplementedError
```

File: cfme/middleware/jdr_report.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        t_mdr.id, t_mdr.queued_at,
+        t_mdr.requesting_user.label('username'),
+        t_mdr.status)\
+        .join(t_mdr, t_mdr.middleware_server_id == t_ms.id)
+    if server:
+        query = query.filter(t_ms.name == server.name)
+    return query
+
+
+class JDRReport():
```
Comment (quarckster) [inline cfme/middleware/jdr_report.py]: **REQUIRED**
Please implement collections http://cfme-tests.readthedocs.io/guides/collections.html

File: cfme/tests/middleware/jdr_report_methods.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+from cfme.utils.wait import wait_for
+from cfme.middleware.jdr_report import JDRReport
```
Comment (quarckster) [inline cfme/tests/middleware/jdr_report_methods.py]: **REQUIRED**
Please follow style guides for imports http://cfme-tests.readthedocs.io/guides/dev_guide.html

File: cfme/middleware/jdr_report.py (hunk 5 above)
Comment (hhovsepy) [reply cfme/middleware/jdr_report.py]: JDRReport itself is not an navigable object, it's tightly coupled to server and is not clickable listed rows for server.

File: cfme/middleware/jdr_report.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+        myreports = JDRReport.reports()
+
+    """
+
+    def __init__(self, username, status, server, appliance=None, **kwargs):
+        self.username = username
+        self.status = status
+        self.server = server
+        self.filename = kwargs['filename'] if 'filename' in kwargs else None
```
Comment (otsuman) [inline cfme/middleware/jdr_report.py]: **REQUIRED**: use ` self.filename = kwargs.get('filename')`

**Comment by hhovsepy** [comment]
@quarckster  @otsuman thank you for review, have squashed all

File: cfme/middleware/jdr_report.py (hunk 3 above)
Comment (hhovsepy) [reply cfme/middleware/jdr_report.py]: have changed it accordingly, not sure why it still there

File: cfme/middleware/jdr_report.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    .. code-block:: python
+
+        myreport = JDRReport(username='admin', status='Ready', server=eap_server)
+        myreport.download()
+
+        myreports = JDRReport.reports()
+
+    """
+
+    def __init__(self, username, status, server, appliance=None, **kwargs):
```
Comment (mfalesni) [inline cfme/middleware/jdr_report.py]: **REQUIRED**: Please, move the `server` argument to the first place.

File: cfme/middleware/jdr_report.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        self.queued_at = kwargs.get('queued_at')
+        self.db_id = kwargs.get('db_id')
+        if 'properties' in kwargs:
+            for prop in kwargs['properties']:
+                # check the properties first, so it will not overwrite core attributes
+                if getattr(self, attributize_string(prop), None) is None:
+                    setattr(self, attributize_string(prop), kwargs['properties'][prop])
+
+    @classmethod
+    def reports(cls, server):
```
Comment (mfalesni) [inline cfme/middleware/jdr_report.py]: This is exactly what are collections good for .... no more classmethods with passing parents around :) These classmethods would live in the collection.

File: cfme/middleware/jdr_report.py (hunk 5 above)
Comment (mfalesni) [reply cfme/middleware/jdr_report.py]: **REQUIRED**: It is not about navigation but about consistent API and dependency passing.

**Comment by hhovsepy** [comment]
Can this PR be reviewed please?, whether collections implementation is correctly done. I wanted to skip making collections in whole middleware provider in scope of this PR.

**Pull request merged**